code/modules/mob/living/carbon/xenomorph/castes/dragon/dragon.dm

```diff
@@ -0,0 +1,184 @@
+	addtimer(CALLBACK(src, PROC_REF(finalize_landing)), 0.5 SECONDS)
+	playsound(src, 'sound/effects/alien/dragon/flying_landing.ogg', 50, TRUE)
+
+/// Finalizes the process of landing by reversing the effects from flying.
+/mob/living/carbon/xenomorph/dragon/proc/finalize_landing()
+	doing_landing_animations = FALSE
+	status_flags = initial(status_flags)
+	resistance_flags = initial(resistance_flags)
+	pass_flags = initial(pass_flags)
+	density = TRUE
+	REMOVE_TRAIT(src, TRAIT_SILENT_FOOTSTEPS, XENO_TRAIT)
+	update_icons(TRUE)
+	update_action_buttons()
+	var/datum/action/ability/activable/xeno/fly/fly_ability = actions_by_path[/datum/action/ability/activable/xeno/fly]
+	fly_ability?.succeed_activate()
+	fly_ability?.add_cooldown()
+
+// Performs various landing effects.
+/mob/living/carbon/xenomorph/dragon/proc/perform_landing_effects(list/turf/impacted_turfs)
+	new /obj/effect/temp_visual/dragon/land(loc)
+	var/damage = 100 * xeno_melee_damage_modifier
+	var/list/obj/vehicle/already_stunned_vehicles = list() // This prevents hitting the main body of a vehicle twice.
+	for(var/turf/impacted_turf AS in impacted_turfs)
+		impacted_turf.Shake(duration = 0.2 SECONDS)
+		for(var/atom/impacted_atom AS in impacted_turf)
+			if(!(impacted_atom.resistance_flags & XENO_DAMAGEABLE))
+				continue
+			if(isxeno(impacted_atom))
+				continue
+			if(isliving(impacted_atom))
+				var/mob/living/impacted_living = impacted_atom
+				if(impacted_living.stat == DEAD)
+					continue
+				impacted_living.take_overall_damage(damage, BRUTE, MELEE, max_limbs = 5, updating_health = TRUE)
+				animate(impacted_living, pixel_z = impacted_living.pixel_z + 8, layer = max(MOB_JUMP_LAYER, impacted_living.layer), time = 0.25 SECONDS, easing = CIRCULAR_EASING|EASE_OUT, flags = ANIMATION_END_NOW|ANIMATION_PARALLEL)
+				animate(pixel_z = impacted_living.pixel_z - 8, layer = impacted_living.layer, time = 0.25 SECONDS, easing = CIRCULAR_EASING|EASE_IN)
+				impacted_living.animation_spin(0.5 SECONDS, 1, impacted_living.dir == WEST ? FALSE : TRUE, anim_flags = ANIMATION_PARALLEL)
+				continue
+			if(!isobj(impacted_atom))
+				continue
+			var/obj/impacted_obj = impacted_atom
+			if(ishitbox(impacted_atom))
+				var/obj/hitbox/impacted_hitbox = impacted_atom
+				var/can_stun = !(impacted_hitbox.root in already_stunned_vehicles)
+				handle_vehicle_effects(impacted_hitbox.root, damage / 3, should_stun = can_stun)
+				if(can_stun)
+					already_stunned_vehicles += impacted_hitbox.root
+				continue
+			if(!isvehicle(impacted_obj))
+				impacted_obj.take_damage(damage, BRUTE, MELEE, blame_mob = src)
+				continue
+			var/can_stun = !(impacted_obj in already_stunned_vehicles)
+			if(ismecha(impacted_obj))
+				handle_vehicle_effects(impacted_obj, damage * 3, 50, should_stun = can_stun)
+			else if(isarmoredvehicle(impacted_obj))
+				handle_vehicle_effects(impacted_obj, damage / 3, should_stun = can_stun)
```

Comment:
see comments about hitboxes yet again

also maybe standardize this into a proccall since you seem to be doing the same thing so often

Reply:
you also seperately check hotboxes so idk why have this actually???